# Wrangle

In [1]:
import pandas as pd

**Data**

- main data: [data.world](https://data.world/makeovermonday/2023w3)
- US abriviation data: [kaggle](https://www.kaggle.com/datasets/agilesifaka/us2letterstatecodecsv/)

In [2]:
migration = pd.read_excel("https://query.data.world/s/al6wfboprlcexlbsk5a44el7ukdqhr?dws=00000")
us = pd.read_csv("US-2-letter-state-code.csv")
migration.head()

Country STATE OR PROVINCE  Year  Inbound  Outbound
0  United States           ALABAMA  2022     0.57      0.43
1  United States           ALABAMA  2021     0.59      0.41
2  United States           ALABAMA  2020     0.61      0.39
3  United States           ALABAMA  2019     0.57      0.43
4  United States           ALABAMA  2018     0.56      0.44

In [3]:
us.head(3)

state_name state_code
0         Alabama         AL
1          Alaska         AK
2  American Samoa         AS

In [4]:
canada_prov = ['Alberta', 'British','Manitoba','New Brunswick','Newfoundland', 'Nova Scotia', 'Northwest Territories', 'Nunavut', 'Ontario', 'Prince Edward Island', 'Quebec','Saskatchewan','Yukon']
ini = ['Alb', 'BC','Man','NB', 'NL', 'NS', 'NWT','NVT', 'ONT', 'PEI', 'Qc', 'Sask', 'YT']

canada = pd.DataFrame({'state_name':canada_prov,
             'state_code':ini})

canada.head(3)

state_name state_code
0    Alberta        Alb
1    British         BC
2   Manitoba        Man

In [5]:
# combine the US and Canada provinces in one dataset
initials = pd.concat([us,canada], axis=0)
initials.state_name = initials.state_name.str.lower()
initials.state_code = initials.state_code.str.upper()
initials.head()

state_name state_code
0         alabama         AL
1          alaska         AK
2  american samoa         AS
3         arizona         AZ
4        arkansas         AR

In [6]:
migration.describe()

Year     Inbound    Outbound
count   610.000000  610.000000  610.000000
mean   2017.500000    0.477689    0.510869
std       2.874639    0.100194    0.102179
min    2013.000000    0.000000    0.000000
25%    2015.000000    0.420000    0.450000
50%    2017.500000    0.490000    0.510000
75%    2020.000000    0.540000    0.570000
max    2022.000000    0.750000    1.000000

In [7]:
migration.shape

(610, 5)

In [8]:
migration.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 610 entries, 0 to 609
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Country            610 non-null    object 
 1   STATE OR PROVINCE  610 non-null    object 
 2   Year               610 non-null    int64  
 3   Inbound            610 non-null    float64
 4   Outbound           610 non-null    float64
dtypes: float64(2), int64(1), object(2)
memory usage: 24.0+ KB


In [9]:
# make columns lower case
migration.columns = migration.columns.str.lower().str.replace(" ", "_")
migration.columns

Index(['country', 'state_or_province', 'year', 'inbound', 'outbound'], dtype='object')

In [10]:
migration.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 610 entries, 0 to 609
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   country            610 non-null    object 
 1   state_or_province  610 non-null    object 
 2   year               610 non-null    int64  
 3   inbound            610 non-null    float64
 4   outbound           610 non-null    float64
dtypes: float64(2), int64(1), object(2)
memory usage: 24.0+ KB


In [11]:
# lower case all states
migration.state_or_province = migration.state_or_province.str.lower()

In [12]:
# merge the state initials on the migration data
migration = migration.merge(initials, left_on='state_or_province', right_on='state_name', how='left')
migration.head()

country state_or_province  year  inbound  outbound state_name  \
0  United States           alabama  2022     0.57      0.43    alabama   
1  United States           alabama  2021     0.59      0.41    alabama   
2  United States           alabama  2020     0.61      0.39    alabama   
3  United States           alabama  2019     0.57      0.43    alabama   
4  United States           alabama  2018     0.56      0.44    alabama   

  state_code  
0         AL  
1         AL  
2         AL  
3         AL  
4         AL

In [13]:
# check for null values
migration.isna().sum()

country               0
state_or_province     0
year                  0
inbound               0
outbound              0
state_name           30
state_code           30
dtype: int64

In [14]:
# make year categorical
migration.year = migration.year.astype("category")

In [15]:
# remove state name column
migration = migration.drop(columns=['state_name'])
migration.head()

country state_or_province  year  inbound  outbound state_code
0  United States           alabama  2022     0.57      0.43         AL
1  United States           alabama  2021     0.59      0.41         AL
2  United States           alabama  2020     0.61      0.39         AL
3  United States           alabama  2019     0.57      0.43         AL
4  United States           alabama  2018     0.56      0.44         AL

In [16]:
# save data
migration.to_csv("Household_goods_migration_patterns.csv")

**Take aways**

- 610 rows and 5 columns
- 0 null counts
- values contained in in/out bound are total shipment percetange

In [16]:
# import tabula
# import pandas as pd

# # Specify the path to your PDF file
# pdf_path = "2022_Migration_Patterns PDF MAP.pdf"

# # Extract tables from the PDF and store them in a list of DataFrames
# tables = tabula.read_pdf(pdf_path, pages="all", multiple_tables=True)

# # Loop through the list of DataFrames and process the tables as needed
# for i, table in enumerate(tables):
#     # Process each table, for example, print it
#     print(f"Table {i + 1}:")
#     print(table)
